presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -34,7 +34,18 @@ default boolean isEnabled(Session session)
         return true;
     }
 
-    Optional<PlanNode> apply(T node, Captures captures, Context context);
+    @Deprecated
+    default Optional<PlanNode> apply(T node, Captures captures, Context context)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    // TODO remove default implementaiton once deprecated apply method got removed
+    default void apply(T node, Captures captures, Context context, Result result)
```

Comment:
I see that `Rule.Result` would would quickly become method proxies class (case class), e.g:

```
CalledAddPlanNodeClass
CalledReplacePlanNodeClass
CalledAddTraitClass
```

Is this really that beneficial? If the only argument for `Rule.Result` is ideological (e.g: all OUT values must be returned as method result value), then I would go for more flexible and extensible design. Java is not a functional language after all.

This way we will have more ways for rule to return plan nodes in the future. It feels Rule.Result is not really the best suited design.

For instance, think of `JoinEnumeration`. It could return enumerations on the go and when it returns enumeration for which cost is acceptable, then the optimizer could break such computations. With `Rule.Result` there is no such possibility.

Generally, there can be an interaction between rule and optimizer (e.g: long running rules could break faster if optimizer tells them to)